app/src/org/commcare/tts/TextToSpeechConverter.kt

```diff
@@ -0,0 +1,191 @@
+            Handler(Looper.getMainLooper()).post {
+                mTTSCallback?.speakFailed()
+            }
+        }
+
+        override fun onStart(utteranceId: String?) { }
+    }
+
+    /**
+     * Adds a listener to register callbacks for different states of Text-To-Speech engine.
+     */
+    fun setListener(listener: TextToSpeechCallback) {
+        mTTSCallback = listener
+    }
+
+    /**
+     * Attempts to speak the specified text.
+     */
+    fun speak(context: Context, text: String) {
+        if (!mInitialized) {
+            initialize(context, text)
+            return
+        }
+        // Handle empty text
+        if (TextUtils.isEmpty(text)) {
+            return
+        }
+        mTextToSpeech?.let { tts ->
+            if (isTextLong(text)) {
+                text.chunked(4000).forEach {
+                    speakInternal(tts, it, TextToSpeech.QUEUE_ADD)
+                }
+            } else {
+                speakInternal(tts, text)
+            }
+        } ?: run {
+            mTTSCallback?.initFailed()
+        }
+    }
+
+    /**
+     * Attempts to stop the TTS.
+     */
+    fun stop() {
+        mTextToSpeech?.let {
+            it.stop()
+        }
+    }
+
+    /**
+     * Attempts to shutdown the TTS engine. No calls should be made to this object after calling this method.
+     * Good to call this from onDestroy().
+     */
+    fun shutDown() {
+        mTextToSpeech?.let {
+            it.shutdown()
+        }
+        mInitialized = false
+    }
+
+    private fun speakInternal(tts: TextToSpeech, text: String, queueMode: Int = TextToSpeech.QUEUE_FLUSH) {
+        val utteranceId = System.currentTimeMillis().toString()
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            tts.speak(text, queueMode, null, utteranceId)
+        } else {
+            val params = HashMap<String, String>()
+            params[TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID] = utteranceId
+            tts.speak(text, queueMode, params)
+        }
+    }
+
+    private fun isTextLong(text: String): Boolean {
+        // TTS can only speak 4000 characters at max at a time.
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            text.length > TextToSpeech.getMaxSpeechInputLength()
+        } else {
+            text.length > MAX_TEXT_LENGTH
+        }
+    }
+
+    /**
+     * Initializes the Text-To-Speech engine
+     */
+    private fun initialize(context: Context, text: String) {
+        mTextToSpeech = TextToSpeech(context, TextToSpeech.OnInitListener { status ->
+            if (status == TextToSpeech.SUCCESS) {
+                // set language and speak
+                mTextToSpeech?.let { tts ->
+                    tts.setOnUtteranceProgressListener(mUtteranceProgressListener)
+                    if (setLocale(tts, LinkedList(listOf(
+                                    Locale(Localization.getCurrentLocale()),
```

Comment:
the locale needs to be reapplied when user changes the language in form and we should also pass the conuntry code from default locale to account for accents.

Reply:
> we should also pass the conuntry code from default locale to account for accents.

How do I get the country codes?

Reply:
You can get the default country code using `Locale.getDefault().getCountry*`

Reply:
Ohh, I thought you were talking about the `Localization.getCurrent` one. 
`Locale.getDefault()` returns a locale with the country code like `en_US`.

Reply:
So If you comprise the current locale as `Locale(Localization.getCurrentLocale(), Locale.getDefault().getCountry())` you should be able to add the country info user is in to the language right ?

Reply:
Ohh, I don't think that makes sense, like if `Localization.getCurrentLocale()` returns Arabic, and `Locale.getDefault().getCountry()` returns US. I'm not sure if it works like that.

Reply:
What I am thinking is if a user in India who has set the device language to Hindi, `Locale.getDefault().getCountry()` should return India as the country and setting that context will improve the TTS by far in comparison to having a US accent for Hindi.   In case of (Arabic, US) the TTS should still return `TextToSpeech.LANG_AVAILABLE` so the behaviour won't get changed in comparison to just (Arabic).

Reply:
ohh ok. I'll test how it works.

Reply:
Wondering if there are any concerns with this approach from the testing you did ?

Reply:
It worked fine for me.